src/condor_dagman/dagman_submit.cpp

```diff
@@ -788,99 +493,109 @@ direct_condor_submit(const Dagman &dm, Job* node,
 		if (submitHash->error_stack()) {
 			submitHash->warn_unused(stderr, "DAGMAN");
 			std::string errstk(submitHash->error_stack()->getFullText());
-			if (!errstk.empty()) {
+			if ( ! errstk.empty()) {
 				debug_printf(DEBUG_QUIET, "Submit warning: %s", errstk.c_str());
 			}
 			submitHash->error_stack()->clear();
 		}
 	}
 
-	if (!tmpDir.Cd2MainDir(errMsg)) {
-		debug_printf(DEBUG_QUIET,
-			"Could not change to original directory: %s\n",
-			errMsg.c_str());
-		success = false;
-	}
-
 	// If the submitHash was only allocated in this function (and not linked to the node) delete it now
-	if (!node->GetSubmitDesc()) {
-		delete submitHash;
+	if ( ! node->GetSubmitDesc()) { delete submitHash; }
+
+	return success;
+}
+
+bool condor_submit(const Dagman &dm, Job* node, CondorID& condorID) {
+	bool success = false;
+	const char* directory = node->GetDirectory();
+	TmpDir tmpDir;
+	std::string errMsg;
+
+	if ( ! tmpDir.Cd2TmpDir(directory, errMsg)) {
+		debug_printf(DEBUG_QUIET, "Could not change to node directory %s: %s\n", directory, errMsg.c_str());
+		return success;
+	}
+
+	DagSubmitMethod method = static_cast<DagSubmitMethod>(dm.options[deep::i::SubmitMethod]);
+	switch (method) {
+		case DagSubmitMethod::CONDOR_SUBMIT: // run condor_submit
+			success = shell_condor_submit(dm, node, condorID);
+			break;
+		case DagSubmitMethod::DIRECT: // direct submit
+			success = direct_condor_submit(dm, node, condorID);
+			break;
+		default:
+			// We have unknown submission method requested so jobs will never be submitted abort
+			debug_printf(DEBUG_NORMAL, "Error: Unknown submit method (%d)\n", (int)method);
+			main_shutdown_rescue(EXIT_ERROR, DAG_STATUS_ERROR);
+			break;
+	}
+
+	if ( ! tmpDir.Cd2MainDir(errMsg)) {
+		debug_printf(DEBUG_QUIET, "Could not change to original directory: %s\n", errMsg.c_str());
+		success = false;
 	}
 
 	return success;
 }
 
-bool send_reschedule(const Dagman & dm)
-{
-	if (dm.options[deep::i::SubmitMethod] == (int)DagSubmitMethod::CONDOR_SUBMIT)
-		return true; // submit already did it
+bool send_reschedule(const Dagman & dm) {
+	DagSubmitMethod method = static_cast<DagSubmitMethod>(dm.options[deep::i::SubmitMethod]);
+	switch (method) {
+		case DagSubmitMethod::CONDOR_SUBMIT: // condor_submit already rescheduled
+			return true;
+		default:
+			break;
+	}
```

Comment:
As an aside, and not for this PR, I'd really prefer us to never switch-on-type, as we do here.  Better to have a virtual base class `DagSubmission` with subclasses for `SpawnedDagSubmission` and `DirectDagSubmission` with methods like `needsReschedule` and `submit`, so there are no switches here in dagman.